### ISSUE TITLE
Post Carousel breakpoint columns not working

### ISSUE THREAD

**Issue by DDevelope** [dudaster/ele-custom-skin#478]
Hi, I'm a Elementor custom skin's pro user. I neep support about the latest update. I use post carousel in my website projects but with the latest release the breakpoint columns not work. I set 3 columns for desktop, 2 for tablet and 1 for mobile but the plugin only work with the 3 columns

**Comment by DDevelope** [comment]
Solved! Replace the js code in the file ecspro.js with this

`var EleCustomSkinSlider = elementorModules.frontend.handlers.Base.extend({
  
  getSkinPrefix: function getSkinPrefix() {
    return 'custom_';
  },  
  getDefaultSettings: function getDefaultSettings() {
    return {
      selectors: {
        carousel: '.swiper-container',
        slideContent: '.swiper-slide'
      }
    }; 
  },

  getDefaultElements: function getDefaultElements() {
    var selectors = this.getSettings('selectors');
    var elements = {
      $carousel: this.$element.find(selectors.carousel)
    };
    elements.$swiperSlides = elements.$carousel.find(selectors.slideContent);
    return elements;
  },
 
  getSlidesCount: function getSlidesCount() {
    return this.elements.$swiperSlides.length;
  },

  getSwiperSettings: function getSwiperSettings() {
    
    var elementSettings = this.getElementSettings(),
      slidesToShow = +elementSettings[this.getSkinPrefix() + 'slides_to_show'] || 3, // this.getElementSettings(this.getSkinPrefix() + 'slides_to_show');
      isSingleSlide = 1 === slidesToShow,
      defaultLGDevicesSlidesCount = isSingleSlide ? 1 : 2,
    //elementorBreakpoints = elementorFrontend.config.breakpoints;
      elementorBreakpoints = elementorFrontend.config.responsive.activeBreakpoints;

    var swiperOptions = {
      slidesPerView: slidesToShow,
      loop: 'yes' === elementSettings[this.getSkinPrefix() + 'infinite'],
      speed: elementSettings[this.getSkinPrefix() + 'speed'],
      handleElementorBreakpoints: true
    };
    swiperOptions.breakpoints = {};
    swiperOptions.breakpoints[elementorBreakpoints.mobile.value] = {
      slidesPerView: +elementSettings[this.getSkinPrefix() + 'slides_to_show_mobile'] || 1,
      slidesPerGroup: +elementSettings[this.getSkinPrefix() + 'slides_to_scroll_mobile'] || 1
    };
    swiperOptions.breakpoints[elementorBreakpoints.tablet.value] = {
      slidesPerView: +elementSettings[this.getSkinPrefix() + 'slides_to_show_tablet'] || defaultLGDevicesSlidesCount,
      slidesPerGroup: +elementSettings[this.getSkinPrefix() + 'slides_to_scroll_tablet'] || defaultLGDevicesSlidesCount
    };
    if (!this.isEdit && 'yes' === elementSettings[this.getSkinPrefix() + 'autoplay']) {
      swiperOptions.autoplay = {
        delay: elementSettings[this.getSkinPrefix() + 'autoplay_speed'],
        disableOnInteraction: !!elementSettings[this.getSkinPrefix() + 'pause_on_hover']
      };
    }

    if (true === swiperOptions.loop) {
      swiperOptions.loopedSlides = this.getSlidesCount();
    }

    if (isSingleSlide) {
      swiperOptions.effect = elementSettings[this.getSkinPrefix() + 'effect'];
    } else {
      swiperOptions.slidesPerGroup = +elementSettings[this.getSkinPrefix() + 'slides_to_scroll'] || defaultLGDevicesSlidesCount;
    }

    if (elementSettings[this.getSkinPrefix() + 'slide_gap'].size) {
      swiperOptions.spaceBetween = elementSettings[this.getSkinPrefix() + 'slide_gap'].size;
    }

    var showArrows = 'arrows' === elementSettings[this.getSkinPrefix() + 'navigation'] || 'both' === elementSettings[this.getSkinPrefix() + 'navigation'],
      showDots = 'dots' === elementSettings[this.getSkinPrefix() + 'navigation'] || 'both' === elementSettings[this.getSkinPrefix() + 'navigation'];

    if (showArrows) {
      swiperOptions.navigation = {
        prevEl: '.elementor-swiper-button-prev',
        nextEl: '.elementor-swiper-button-next'
      };
    }

    if (showDots) {
      swiperOptions.pagination = {
        el: '.swiper-pagination',
        type: 'bullets',
        clickable: true
      };
    }
console.log(swiperOptions);
    return swiperOptions;
  },

  updateSpaceBetween: function updateSpaceBetween() {
    this.swiper.params.spaceBetween = this.getElementSettings(this.getSkinPrefix() + 'slide_gap') || 0;
    this.swiper.update();
  },
  isSliderEnabled: function isSliderEnabled() {
    return !!this.getElementSettings(this.getSkinPrefix() + 'post_slider');
  },
  run: function run(){
    if(!this.isSliderEnabled()) return;
    
    if (!this.elements.$carousel.length) {
      return;
    }
//here we call the swiper
    if ( 'undefined' === typeof Swiper ) {
      const asyncSwiper = elementorFrontend.utils.swiper;

      new asyncSwiper( this.elements.$carousel, this.getSwiperSettings() ).then( ( newSwiperInstance ) => { 
        this.swiper = newSwiperInstance;
      } );
    } else {	 
      this.swiper = new Swiper( this.elements.$carousel, this.getSwiperSettings() );
    } 
    
    //add reinitilze for swiper
    ECScarousel = this.elements.$carousel;
    ECSswiperSettings = this.getSwiperSettings();
    
    ECS_add_action("swiper", function(){
      ECSswiper = new Swiper(ECScarousel, ECSswiperSettings);
    });
    
    
  },
  onInit: function onInit() {
    
    elementorModules.frontend.handlers.Base.prototype.onInit.apply(this, arguments);
    this.run();
  },

  onElementChange: function onElementChange(propertyName) {
 //    console.log(this.swiper);
    if (0 === propertyName.indexOf(this.getSkinPrefix() + 'slide_gap')) {
      this.updateSpaceBetween();
    }
  },

  onEditSettingsChange: function onEditSettingsChange(propertyName) { // here you need to refresh stuff when it
    if(!this.isSliderEnabled()) return;
    if ('activeItemIndex' === propertyName) {
      this.swiper.slideToLoop(this.getEditSettings('activeItemIndex') - 1);
    }
  }
});

var EleCustomSkinSliderArchive = EleCustomSkinSlider.extend({
  getSkinPrefix: function getSkinPrefix() {
    return 'archive_custom_';
  }
});

//now let's see if we can call it
jQuery(window).on('elementor/frontend/init', () => {

  const addHandler = ($element) => {
    elementorFrontend.elementsHandler.addHandler(EleCustomSkinSlider, {
      $element,
    });
  };


  elementorFrontend.hooks.addAction('frontend/element_ready/posts.custom', addHandler);
  
    const addHandlerArchive = ($element) => {
    elementorFrontend.elementsHandler.addHandler(EleCustomSkinSliderArchive, {
      $element,
    });
  };


  elementorFrontend.hooks.addAction('frontend/element_ready/archive-posts.archive_custom', addHandlerArchive);

});


function ECS_reset_masonry(element){
 element.children(".ecs-post-loop").css('margin-top', '');
}

var EleCustomSkinMasonry = elementorModules.frontend.handlers.Base.extend({
  getSkinPrefix: function getSkinPrefix() {
    return 'custom_';
  },

  bindEvents: function bindEvents() {
    var cid = this.getModelCID();

    elementorFrontend.addListenerOnce(cid, 'resize', this.onWindowResize);
  },

  getClosureMethodsNames: function getClosureMethodsNames() {
    return elementorModules.frontend.handlers.Base.prototype.getClosureMethodsNames.apply(this, arguments).concat(['fitImages', 'onWindowResize', 'runMasonry']);
  },

  getDefaultSettings: function getDefaultSettings() {
    return {
      classes: {
        fitHeight: 'elementor-fit-height',
        hasItemRatio: 'elementor-has-item-ratio'
      },
      selectors: {
        postsContainer: '.elementor-posts-container',
        post: '.elementor-post',
        postThumbnail: '.elementor-post__thumbnail',
        postThumbnailImage: '.elementor-post__thumbnail img'
      }
    };
  },

  getDefaultElements: function getDefaultElements() {
    var selectors = this.getSettings('selectors');

    return {
      $postsContainer: this.$element.find(selectors.postsContainer),
      $posts: this.$element.find(selectors.post)
    };
  },

  fitImage: function fitImage($post) {
    var settings = this.getSettings(),
      $imageParent = $post.find(settings.selectors.postThumbnail),
      $image = $imageParent.find('img'),
      image = $image[0];

    if (!image) {
      return;
    }

    var imageParentRatio = $imageParent.outerHeight() / $imageParent.outerWidth(),
      imageRatio = image.naturalHeight / image.naturalWidth;

    $imageParent.toggleClass(settings.classes.fitHeight, imageRatio < imageParentRatio);
  },

  fitImages: function fitImages() {
    var $ = jQuery,
      self = this,
      itemRatio = getComputedStyle(this.$element[0], ':after').content,
      settings = this.getSettings();

    this.elements.$postsContainer.toggleClass(settings.classes.hasItemRatio, !!itemRatio.match(/\d/));

    if (self.isMasonryEnabled()) {
      return;
    }

    this.elements.$posts.each(function() {
      var $post = $(this),
        $image = $post.find(settings.selectors.postThumbnailImage);

      self.fitImage($post);

      $image.on('load', function() {
        self.fitImage($post);
      });
    });
  },

  setColsCountSettings: function setColsCountSettings() {
    var currentDeviceMode = elementorFrontend.getCurrentDeviceMode(),
      settings = this.getElementSettings(),
      skinPrefix = this.getSkinPrefix(),
      colsCount;

    switch (currentDeviceMode) {
      case 'mobile':
        colsCount = settings[skinPrefix + 'columns_mobile'];
        break;
      case 'tablet':
        colsCount = settings[skinPrefix + 'columns_tablet'];
        break;
      default:
        colsCount = settings[skinPrefix + 'columns'];
    }

    this.setSettings('colsCount', colsCount);
  },

  isMasonryEnabled: function isMasonryEnabled() {
    return !!this.getElementSettings(this.getSkinPrefix() + 'masonrys');
  },

  initMasonry: function initMasonry() {
    imagesLoaded(this.elements.$posts, this.runMasonry);
  },

  runMasonry: function runMasonry() {
    var elements = this.elements;

    elements.$posts.css({
      marginTop: '',
      transitionDuration: ''
    });

    this.setColsCountSettings();

    var colsCount = this.getSettings('colsCount'),
      hasMasonry = this.isMasonryEnabled();

    elements.$postsContainer.toggleClass('elementor-posts-masonry', hasMasonry);

    if (!hasMasonry) {
      elements.$postsContainer.height('');

      return;
    }

    /* The `verticalSpaceBetween` variable is setup in a way that supports older versions of the portfolio widget */

    var verticalSpaceBetween = this.getElementSettings(this.getSkinPrefix() + 'row_gap.size');

    if ('' === this.getSkinPrefix() && '' === verticalSpaceBetween) {
      verticalSpaceBetween = this.getElementSettings(this.getSkinPrefix() + 'item_gap.size');
    }

    var masonry = new elementorModules.utils.Masonry({
      container: elements.$postsContainer,
      items: elements.$postsContainer.children( ".ecs-post-loop" ),//elements.$posts.filter(':visible'),
      columnsCount: this.getSettings('colsCount'),
      verticalSpaceBetween: verticalSpaceBetween
    });
    masonry.run();
    
    ECS_Columns_Count = this.getSettings('colsCount');
    ECS_add_action("ajax", function(){//add reinitilze to ajax
      ECS_reset_masonry(elements.$postsContainer);
      mymasonry = new elementorModules.utils.Masonry({
              container: elements.$postsContainer,
              items: elements.$postsContainer.children( ".ecs-post-loop" ),
              columnsCount: ECS_Columns_Count,
              verticalSpaceBetween: verticalSpaceBetween
            });
            mymasonry.run();

     });

  },

  run: function run() {
    // For slow browsers
    setTimeout(this.fitImages, 0);

    this.initMasonry();
  },

  onInit: function onInit() {

    elementorModules.frontend.handlers.Base.prototype.onInit.apply(this, arguments);

    this.bindEvents();

    this.run();
  },

  onWindowResize: function onWindowResize() {
    this.fitImages();
	  ECS_reset_masonry(this.elements.$postsContainer);
    this.runMasonry();
  },

  onElementChange: function onElementChange() {
    this.fitImages();

    setTimeout(this.runMasonry);
  }
});

var EleCustomSkinMasonryArchive = EleCustomSkinMasonry.extend({
  getSkinPrefix: function getSkinPrefix() {
    return 'archive_custom_';
  }
});
//now let's see if we can call it
jQuery(window).on('elementor/frontend/init', () => {

  const addHandler = ($element) => {
    elementorFrontend.elementsHandler.addHandler(EleCustomSkinMasonry, {
      $element,
    });
  };

  elementorFrontend.hooks.addAction('frontend/element_ready/posts.custom', addHandler);
  
  const addHandlerArchive = ($element) => {
    elementorFrontend.elementsHandler.addHandler(EleCustomSkinMasonryArchive, {
      $element,
    });
  };


  elementorFrontend.hooks.addAction('frontend/element_ready/archive-posts.archive_custom', addHandlerArchive);

});

//make item clickable


function EleCustomSkinItemLink() {
  // jQuery(".ecs-link-wrapper").attr("tabindex","0");
  jQuery(".ecs-link-wrapper")
    .on("click", function (event) {
      loc = jQuery(this).attr("data-href");
      aTarget = jQuery(this).attr("data-target");
      if (
        !jQuery(event.target).closest(
          ".elementor-swiper-button, .swiper-pagination, .ecs-ignore-link"
        ).length
      ) {
        if (aTarget == "_blank" || event.ctrlKey || event.metaKey) {
          window.open(loc);
        } else {
          window.location = loc;
        }
      }
    
    })
    .on("click", ".ecs-ignore-link", function (e) {
      if(!jQuery(e.target).find('.elementor-button-wrapper')){
        e.stopPropagation();
      }
    });
}

/*function EleCustomSkinItemLink(){
   // jQuery(".ecs-link-wrapper").attr("tabindex","0");
    jQuery(".ecs-link-wrapper").click(function(event) {
     loc = jQuery(this).attr("data-href");
     aTarget = jQuery(this).attr("data-target");
     if (!jQuery(event.target).closest('.elementor-swiper-button, .swiper-pagination, .ecs-ignore-link').length) {
       if(aTarget == "_blank" || event.ctrlKey || event.metaKey){
            window.open(loc);
        } else {
            window.location = loc;
        }
      }
      return false;
  }); 
}*/

jQuery( document ).ready(function() {
  EleCustomSkinItemLink();

  ECS_add_action("ajax", function(){EleCustomSkinItemLink()});

});

jQuery.fn.inView = function(){
    if(!this.length) 
        return false;
    var rect = this.get(0).getBoundingClientRect();
    var loadOffset=0;
    return (
        rect.top >= 0 &&
        rect.left >= 0 &&
        rect.bottom - loadOffset <= (window.innerHeight || document.documentElement.clientHeight) &&
        rect.right <= (window.innerWidth || document.documentElement.clientWidth)
    );

};

function ECScheckInView(){
    jQuery('.ecs-lazyload').each(function(){
        if (jQuery(this).inView()){
          id = jQuery(this).attr("data-targetid");
          //console.log(id);
          jQuery(this).removeClass("animation-hidden");
          ECS_load_next_page(id);
        }
    });
}

jQuery(function($){
	if(!$( "body" ).hasClass( "elementor-editor-active" ) && !$( "body" ).hasClass( "elementor-editor-preview" ) ) {
    ECScheckInView();
    $(window).scroll(function(){
      if(canBeLoaded == true ){
          ECScheckInView();
        }
  	});
  }
});
`

**Comment by dudaster** [comment]
Thank you DDevelope for the heads up. It is a bug and it is fixed and it will be available with the next version!

**Comment by ffilp** [comment]
The same problem for me, I tried the 3.1.0 Beta 2-version without any luck.

My temporary no-coding fix until the next version is to create a duplicate of the Post Carousel with 1 column, only visible on mobile devices.

**Comment by dionisis17** [comment]
I got the same problem. I am elementor and custom skin PRO user and I have this issue in mobile and tablet mode.
My plugins are all updated to latest version.
![ecskin1](https://user-images.githubusercontent.com/53856507/116048406-cf103200-a67d-11eb-9315-7815fbffc058.png)
![ecskin2](https://user-images.githubusercontent.com/53856507/116048412-d0d9f580-a67d-11eb-9645-db0cfff05045.png)

**Comment by dudaster** [comment]
Latest version is 3.1.0 for both. Make sure you have them!

**Comment by dionisis17** [comment]
> 
> 
> Latest version is 3.1.0 for both. Make sure you have them!

I do
![issue_ecs](https://user-images.githubusercontent.com/53856507/116052795-a50d3e80-a682-11eb-845c-e367ae278cde.png)

**Comment by dionisis17** [comment]
Elementor released 3.2.2. I do not know if that helped but the tablet settings **now work**. The phone settings still doesn't !
Could you please check it? 
Thank you!

**Comment by dudaster** [comment]
I will investigate!

**Comment by dudaster** [comment]
At my end is it working. I will need the url of your website to inspect element to see what is going on there.

**Comment by dionisis17** [comment]
Alright, sure.
Log in here <https://ishop.com.gr/>with
demo
demo

and go to the home page by clicking the logo on the top left corner.



Στις Τετ, 28 Απρ 2021 στις 11:24 π.μ., ο/η Liviu Duda <
***@***.***> έγραψε:

**Comment by dudaster** [comment]
for me works like this: 4 -desktop, 3- tablet, 1- mobile (number of columns)

**Comment by dionisis17** [comment]
yes, sorry I want in phone to show 2 columns and it doesn't work. Let me
change it.

Στις Τετ, 28 Απρ 2021 στις 11:47 π.μ., ο/η Liviu Duda <
***@***.***> έγραψε:

**Comment by dionisis17** [comment]
![Firefox_Screenshot_2021-04-28T08-51-55 052Z](https://user-images.githubusercontent.com/53856507/116375853-585b6c00-a818-11eb-940f-08e544ebcea1.png)
I changed it. Inspect now

Στις Τετ, 28 Απρ 2021 στις 11:49 π.μ., ο/η Διονύσης Σωτηρίου <
***@***.***> έγραψε:

**Comment by dudaster** [comment]
this is a CSS conflict issue with OceanWP template. In fact the problem is this code:

    .col:not(.slick-slide),.blog-entry {
        width: 100% !important;
        margin-bottom: 25px;
    }

And OceanWP adds this class to the items of the posts widget:

`<article id="post-1806" class="elementor-post elementor-grid-item ecs-post-loop swiper-slide post-1806 product type-product status-publish has-post-thumbnail product_cat-laptop entry has-media first instock shipping-taxable purchasable product-type-simple has-product-nav col span_1_of_3 owp-content-center owp-thumbs-layout-horizontal owp-btn-normal owp-tabs-layout-horizontal swiper-slide-duplicate swiper-slide-duplicate-prev" data-swiper-slide-index="0" style="width: 170px; margin-right: 15px;">`

and this is how the code looks without ocean wp:

`<article id="post-80" class="elementor-post elementor-grid-item ecs-post-loop swiper-slide post-80 post type-post status-publish format-standard has-post-thumbnail hentry category-category-0 category-category-2 category-category-3 category-category-5" data-swiper-slide-index="2" style="width: 160px; margin-right: 15px;">`

as you can see the Swiper Slider sets the width of the article with javascript inside article tag.

**Comment by dudaster** [comment]
Solution:
Try this code inside a html widget inside footer template

```
<script>
jQuery(".swiper-slide").addClass("slick-slide");
</script>
```

**Comment by dionisis17** [comment]
I tried it. It works!. Thank you very much! I appreciate it!
But the strange thing is that the result is not showing right in the backend (in elementor editing). In frontend showing right though.

**Comment by dudaster** [comment]
Fixed with 3.1.0
